module/VuFind/src/VuFind/ILS/Driver/Folio.php

```diff
@@ -251,9 +251,16 @@ public function preRequest(\Laminas\Http\Headers $headers, $params)
     protected function renewTenantToken()
     {
         $this->token = null;
+        $password = $this->config['API']['password'];
```

Comment:
I think it might be useful to create a \VuFind\Config\Feature\SecretTrait.php file containing something like:

```
namespace VuFind\Config\Feature;

use Laminas\Config\Config;

trait SecretTrait
{
    function getSecretFromConfigOrSecretFile(Config|array|null $config, string $key): ?string
    {
        if (!$config) {
            return null;
        }
        if ($config instanceof Config) {
            $config = $config->toArray();
        }
        return isset($config[$key . '_file'])
            ? trim(file_get_contents($config[$key . '_file']))
            : $config[$key] ?? null
    }
}
```

That's untested code, but if I wrote it correctly, it should allow you to pass in an array or a Config object and a key, and then return either the raw value or a file-based override. I think using this trait in all of the affected classes should make things a little more readable. It also gives us a single integration point if we wanted to add additional mechanisms for secret retrieval in the future. What do you think?

Reply:
Yes, you are right, I didn't thought of making a trait to remove the redundancy, thanks!